siff-dr-router/router-netfilter.go

```diff
@@ -1,20 +1,73 @@
 package main
 
 import (
+	"bytes"
 	"log"
 
 	"github.com/ThomasJClark/cs4516project/pkg/go-netfilter-queue"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
 )
 
+func Serialize(ipLayer *layers.IPv4) ([]byte, error) {
+	/*Write the IPv4 header into a gopacket buffer*/
+	buf := gopacket.NewSerializeBuffer()
+	err := ipLayer.SerializeTo(buf, gopacket.SerializeOptions{FixLengths: false, ComputeChecksums: true})
+	if err != nil {
+		return nil, err
+	}
+
+	/*Write the gopacket buffer and the payload into a byte buffer, concatenating
+	  the entire packet together.*/
+	var buf2 bytes.Buffer
+	buf2.Write(buf.Bytes())
+	buf2.Write(ipLayer.Payload)
+
+	return buf2.Bytes(), nil
+}
+
 func processPackets() {
+	log.Println("Starting router queue")
 	nfq, err := netfilter.NewNFQueue(0, 100000, 0xffff)
 	if err != nil {
 		log.Fatal(err)
 	}
 
 	for packet := range nfq.GetPackets() {
+		log.Println("Got packet")
+
+		if isSiff(&packet) {
+			capability := calcCapability(&packet)
+			if isExp(&packet) {
+				log.Println("Got exp packet")
+				addCapability(&packet, capability)
+			} else {
+				capabilities := getCapabilities(&packet)
+				if len(capabilities) < 1 || capability != getCapabilities(&packet)[0] {
+					log.Println("Capability mismatch")
+					packet.SetVerdict(netfilter.NF_DROP)
+				}
+			}
+			shiftCapability(&packet)
```

Comment:
I don't think the router is supposed to do a shift if its an exp packet. See http://repository.cmu.edu/cgi/viewcontent.cgi?article=1035&context=ece page 6